apiserver/facades/agent/storageprovisioner/storageprovisioner.go

```diff
@@ -1700,3 +1704,46 @@ func (s *StorageProvisionerAPIv4) RemoveAttachment(ctx context.Context, args par
 	}
 	return results, nil
 }
+
+// SetStatus sets the status of each given storage artefact.
+func (s *StorageProvisionerAPIv4) SetStatus(ctx context.Context, args params.SetStatus) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Entities)),
+	}
+	if len(args.Entities) == 0 {
+		return result, nil
+	}
+	canModify, err := s.getStorageEntityAuthFunc(ctx)
+	if err != nil {
+		return params.ErrorResults{}, err
+	}
+	now := s.clock.Now()
+	for i, arg := range args.Entities {
+		tag, err := names.ParseTag(arg.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if !canModify(tag) {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		sInfo := status.StatusInfo{
+			Status:  status.Status(arg.Status),
+			Message: arg.Info,
+			Data:    arg.Data,
+			Since:   &now,
+		}
+		var statusErr error
+		switch tag := tag.(type) {
+		case names.FilesystemTag:
+			statusErr = s.storageStatusService.SetVolumeStatus(ctx, tag.Id(), sInfo)
+		case names.VolumeTag:
+			statusErr = s.storageStatusService.SetVolumeStatus(ctx, tag.Id(), sInfo)
+		default:
+			statusErr = apiservererrors.ErrPerm
+		}
+		result.Results[i].Error = apiservererrors.ServerError(statusErr)
```

Comment:
We need to handle the errors of the service.

Reply:
true, done